apps/gdalalg_raster_rgb_to_palette.cpp

```diff
@@ -0,0 +1,323 @@
+
+#include "cpl_string.h"
+#include "gdal_alg.h"
+#include "gdal_priv.h"
+#include "commonutils.h"
+
+#include <array>
+
+//! @cond Doxygen_Suppress
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+/************************************************************************/
+/*                    GDALRasterRGBToPaletteAlgorithm()                 */
+/************************************************************************/
+
+GDALRasterRGBToPaletteAlgorithm::GDALRasterRGBToPaletteAlgorithm()
+    : GDALAlgorithm(NAME, DESCRIPTION, HELP_URL)
+{
+    AddProgressArg();
+
+    AddOpenOptionsArg(&m_openOptions);
+    AddInputFormatsArg(&m_inputFormats)
+        .AddMetadataItem(GAAMDI_REQUIRED_CAPABILITIES, {GDAL_DCAP_RASTER});
+    AddInputDatasetArg(&m_inputDataset, GDAL_OF_RASTER);
+
+    AddOutputDatasetArg(&m_outputDataset, GDAL_OF_RASTER);
+    AddOutputFormatArg(&m_format, /* bStreamAllowed = */ false,
+                       /* bGDALGAllowed = */ false)
+        .AddMetadataItem(GAAMDI_REQUIRED_CAPABILITIES,
+                         {GDAL_DCAP_RASTER, GDAL_DCAP_CREATE});
+    AddCreationOptionsArg(&m_creationOptions);
+    AddOverwriteArg(&m_overwrite);
+
+    AddArg("color-count", 0,
+           _("Select the number of colors in the generated color table"),
+           &m_colorCount)
+        .SetDefault(m_colorCount)
+        .SetMinValueIncluded(2)
+        .SetMaxValueIncluded(256);
+    AddArg("color-map", 0, _("Color map filename"), &m_colorMap);
+}
+
+/************************************************************************/
+/*                GDALRasterRGBToPaletteAlgorithm::RunImpl()            */
+/************************************************************************/
+
+bool GDALRasterRGBToPaletteAlgorithm::RunImpl(GDALProgressFunc pfnProgress,
+                                              void *pProgressData)
+{
+    auto poSrcDS = m_inputDataset.GetDatasetRef();
+    CPLAssert(poSrcDS);
+
+    if (m_format.empty())
+    {
+        m_format = GetOutputDriverForRaster(m_outputDataset.GetName().c_str());
+        if (m_format.empty())
+        {
+            ReportError(CE_Failure, CPLE_AppDefined,
+                        "Cannot guess output driver from output filename");
+            return false;
+        }
+    }
+    if (EQUAL(m_format.c_str(), "JPEG") || EQUAL(m_format.c_str(), "WEBP"))
+    {
+        ReportError(CE_Failure, CPLE_NotSupported,
+                    "Format %s does not support color tables",
+                    CPLString(m_format).toupper().c_str());
+        return false;
+    }
+
+    const int nSrcBandCount = poSrcDS->GetRasterCount();
+    if (nSrcBandCount < 3)
+    {
+        ReportError(CE_Failure, CPLE_NotSupported,
+                    "Input dataset must have at least 3 bands");
+        return false;
+    }
+    else if (nSrcBandCount >= 4)
+    {
+        ReportError(
+            CE_Warning, CPLE_AppDefined,
+            "Only R,G,B bands of input dataset will be taken into account");
```

Comment:
if alpha is not taken, shouldn't this check  be >= 3 ?

Reply:
we want to warn if there is an alpha band, hence >= 4